feature (protocols): force loading step of group strategy

gchoqueux opened the pull request:
## Description
Force loading step of group strategy and don't use parent texture, if level is step of group

## Motivation and Context
When node use parent's texture elevation which is distant in the hierarchy,
OBB's size can be too big and it can be subdivided unnecessarily and the joints between the tiles are too different (see screenshot)


## Screenshots (if appropriate)
![capture du 2017-10-19 11-27-38](https://user-images.githubusercontent.com/11291849/31764912-ec3f12ca-b4c2-11e7-8146-06734f182011.png)

**New steps to reproduce**
* Start localhost/index.htlm with only ortho layer (size `viewport 1100x1100`)
* when globe is loaded call in console : 
`globeView.controls.setCameraTargetGeoPositionAdvanced({ longitude:  6.886991921801097, latitude: 45.87929727669077, range: 2000});`

**Result in master, ~45 tiles of level 17  and ~420 wmts requests**
![submaster](https://user-images.githubusercontent.com/11291849/34829686-cb9ec9f0-f6e1-11e7-8c9f-5c4aa8d6a9b4.png)

**Result in PR, ~40 tiles of level 16 and ~340 wmts requests**
![downsub](https://user-images.githubusercontent.com/11291849/34829761-0cb0a83c-f6e2-11e7-9445-d67dd0cf0fc3.png)

the error comes from inheritance of textures elevations with a large geometric error and which induces more subdivisions

peppsac commented:
I'd like to try to reproduce the issue here.
Can you wrtie specific steps to reproduce it (for instance camera position/rotation of the above screenshot) ?

zarov commented:
@gchoqueux could you provide a camera position / instructions where the problem occurs, so we can see where the fix operates ?

gchoqueux commented:
~~it isn't the same consequences as the subdivisions. 
But the heritage altitudes (low zoom) also causes culling bug. because bounding box isn't representative (the altitudes with a low zoom have a large error)~~

This bug isn't the same issue exposed in this PR.
It comes from a calculation error of the bounding box (caused by the sub-sampling of searches in the elevation data)

![culling3](https://user-images.githubusercontent.com/11291849/33887290-9666dd54-df49-11e7-85bf-6f5d890bfb87.png)

**Steps to reproduce in globe**
```js
// 1 execute in console
globeView.controls.setCameraTargetGeoPositionAdvanced({ longitude:  6.886991921801097, latitude: 45.87929727669077, range: 2000});
// wait ending and  execute in console
globeView.controls.setCameraTargetGeoPositionAdvanced({ longitude:  6.886991921801097, latitude: 45.87929727669077, tilt: 70, heading: 180,  range: 1500});
```

gchoqueux commented:
@iTowns/reviewers branch ready and updated

peppsac commented on src/Process/SubdivisionControl.js:
```diff
@@ -4,9 +4,25 @@ export default {
+        context.forceElevation.sort((a, b) => a - b);
+    },
     hasEnoughTexturesToSubdivide: (context, layer, node) => {
+        if (context.forceElevation.length) {
```
Can you document what are the next 5 lines doing, please?

peppsac replied:
My guess is: you prevent subdivision of a node if its elevation level is < to max(valid group levels <= node.level)?

So you need to explain what you do and why you need to do it (maybe with an example?)

peppsac commented on examples/layers/JSONLayers/WORLD_DTM.json:
```diff
@@ -8,7 +8,8 @@
         "options": {
-            "groups": [3, 7, 9]
+            "groups": [3, 7, 9],
+            "forceLoadingStep": true
```
It seems you added `forceLoadingStep = true` for all layers using the Group strategy.
Could it still make sense to use the Group stragegy without `forceLoadingStep`?

gchoqueux replied:
completely agree

peppsac commented on src/Process/SubdivisionControl.js:
```diff
@@ -4,9 +4,25 @@ export default {
 
+        context.forceElevation = [];
+        for (const e of context.elevationLayers) {
+            if (e.updateStrategy.type == 1 && e.updateStrategy.options.forceLoadingStep && e.updateStrategy.options.groups) {
```
(replace `1` with `STRATEGY_GROUP`)

peppsac commented on src/Process/SubdivisionControl.js:
```diff
@@ -4,9 +4,25 @@ export default {
+                context.forceElevation = context.forceElevation.concat(e.updateStrategy.options.groups);
+            }
+        }
+        context.forceElevation.sort((a, b) => a - b);
```
Can you comment / use a more descriptive variable name?
I think `forceElevation` is a sorted array of all defined valid level for elevation layers using the group strategy. Is that correct?

peppsac reviewed (changes requested):
Thanks for the repro steps.

Your fix needs a bit more comments.
Also did you consider adding a unit test for this issue?

gchoqueux commented on src/Main.js:
```diff
@@ -24,3 +24,6 @@ export { default as GeoJSON2Features } from './Renderer/ThreeExtended/GeoJSON2Fe
 export { default as DEMUtils } from './utils/DEMUtils';
+// Export for test
+// Question: how export TileMesh only for the test?
+export { default as TileMesh } from './Core/TileMesh';
```
How export `TileMesh` only for the test?
I modify `TileMesh` function ([here](https://github.com/iTowns/itowns/pull/538/files#diff-7c875b4c6fb106f12afe62df279296c5R7)) only for the test to not pollute the code and use.

autra replied:
You can import TileMesh in tests without adding it in Main.js, you just need the correct path, see  #594 for instance (and the existing unit tests).

autra replied:
And in #594 you have also some functions exported only for testing.

gchoqueux replied:
@autra thanks!

gchoqueux commented:
branch updated

peppsac commented on src/Process/SubdivisionControl.js:
```diff
@@ -1,12 +1,33 @@
+        // Prevent to subdivise the node if the current elevation level
+        // is below the target level in group strategy.
+        // we must avoid a tile, with level 20, inherits a level 3 elevation texture.
+        // The induced geometric error is much too large and distorts the SSE
```
I fail to see why this is specific to the `STRATEGY_GROUP`?
Basically you want to prevent subdivision if `node.level - elevationLevel` is too big.

And IMHO, the bug should be fixed in the SSE logic (https://github.com/iTowns/itowns/blob/master/src/Process/LayeredMaterialNodeProcessing.js#L60 is another place that got patched for the same SSE logic issue).

gchoqueux replied:
> I fail to see why this is specific to the STRATEGY_GROUP?

indeed, we can extend this logic to all strategies.

> And IMHO, the bug should be fixed in the SSE logic (https://github.com/iTowns/itowns/blob/master/src/Process/LayeredMaterialNodeProcessing.js#L60 is another place that got patched for the same SSE logic issue).

On the other hand, if this logic is placed at this point, the tile will be displayed first without any elevation texture, it will be a regression..

gchoqueux committed:
fix(core): Prevent to subdivide for poor elevation level

we must avoid a tile, with level 20, inherits a level 3 elevation texture.
The induced geometric error is much too large and distorts the SSE

gchoqueux committed:
test(core): should subdivide like expected and prevent to subdivide for poor elevation level

gchoqueux commented:
@peppsac  
I extend this logic to all strategy and moved it SSE processing

gchoqueux commented:
This PR is updated

gchoqueux commented:
**New steps to reproduce** in title

Bug related here https://github.com/iTowns/itowns/pull/538#issuecomment-351053532
doesn't correspond to the modification of this PR

peppsac reviewed (approved):
Ok, let's merge this PR: it's an acceptalbe workaround for a bug in the SSE mechanism.
